### PR TITLE
Bail out of Patterns with too long placeables

### LINKED ISSUE #388: Unskip the Billion Laughs test
Let's fix it.

### PR DESCRIPTION
**Pull request by stasm**
Fix #388.

### CODE REVIEW

File: fluent/src/resolver.js [hunk 1]
```diff
@@ -289,11 +289,12 @@ export function resolveComplexPattern(scope, ptn) {
           `(${part.length}, max allowed is ${MAX_PLACEABLE_LENGTH})`
         )
       );
-      result.push(part.slice(MAX_PLACEABLE_LENGTH));
-    } else {
-      result.push(part);
+      scope.dirty.delete(ptn);
+      return new FluentNone();
```
Comment (stasm) [inline fluent/src/resolver.js]: This isn't perfect, but does provide some level of protection. The problem this approach still suffers from is that `FluentNone` will then `toString`-ed to `{???}` many many times over, depending on how many patterns fail in references up the nesting tree. This also means that the number of accumulated errors may be huge.

I filed https://github.com/projectfluent/fluent/issues/277 to continue the discussion in the spec.

**Comment by Pike** [comment]
I wonder if we should `throw` to bail out of the attack. But that puts us back into the square of having to `try-catch` every call? Well, not every, we could do that in `formatPattern`, and that would only do it for complex patterns.

You had concerns around `try-catch` and perf?

I'm having second thoughts on this, in particular around question like how this would be implemented with `formatToParts` where we don't know the length. Or in Rust, where we can't just throw an exception and clean things up at the top-level.

Is this specific implementation practically better than what we have? Maybe it's not worth to tweak the status quo right now, and modify the behavior once we have something spec'ed?

**Comment by stasm** [comment]
> You had concerns around try-catch and perf?

Yes, but I'll need to create a benchmark to verify this. I'll report in #364 or #390, or both.

> Is this specific implementation practically better than what we have? Maybe it's not worth to tweak the status quo right now, and modify the behavior once we have something spec'ed?

Yes, my rationale for opening this PR was along the same lines. This isn't perfect, but it is better that what we currently do on master.

**Comment by stasm** [comment]
I changed the approach to treating the error as a _fatal_ one, which causes the resolver to instantly bail out of the affected pattern. This provides a much better CPU protection in deeply nested scenarios. I think we could even consider increasing `MAX_PLACEABLE_LENGTH`, but I left it for another time.

**Comment by stasm** [comment]
Using `try...catch` has no impact on our benchmarks.

**Comment by stasm** [comment]
Thanks for the review and the suggestion to bail out in this particular case!

**Pull request merged**